{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Module *requests*\n",
    "\n",
    "Pour interroger un API (REST). Un petit survol disponible [ici](https://www.w3schools.com/python/module_requests.asp)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "url    = 'https://jsonplaceholder.typicode.com/posts'  # Url utile pour faire des tests\n",
    "params = {'userId':1}                                  # Paramètres sous forme de dictionnaire, ici id=1\n",
    "req    = requests.get(url, params)                     # Ici méthode get(), mais on peut aussi utiliser post()\n",
    "\n",
    "\n",
    "# Méthode json() si le format retourné est un json/liste de json.\n",
    "# Pour le format texte on utilise 'req.text', sans parenthèses.\n",
    "for i in req.json(): print(i)\n",
    "req.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# API *Entrez* du NCBI\n",
    "\n",
    "Je recommande fortement d'aller voir:\n",
    "\n",
    "- [The 9 E-utilities and Associated Parameters](https://dataguide.nlm.nih.gov/eutilities/utilities.html)\n",
    "- [A General Introduction to the E-utilities](https://www.ncbi.nlm.nih.gov/books/NBK25497/)\n",
    "- [The E-utilities In-Depth: Parameters, Syntax and More](https://www.ncbi.nlm.nih.gov/books/NBK25499/)\n",
    "\n",
    "Liste des bases de données disponibles [ici](https://www.ncbi.nlm.nih.gov/books/NBK25497/table/chapter2.T._entrez_unique_identifiers_ui/?report=objectonly)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ESearch\n",
    "\n",
    "On cherche une base de donnée (ici la bd 'gene') à l'aide de mots-clés. Retourne, entre autres, la liste des identifiants des gènes correspondants. Cet identifiant va nous permettre d'aller chercher plus d'infos sur les gènes correspondant plus tard. Plus de détail [ici](https://www.ncbi.nlm.nih.gov/books/NBK25499/#_chapter4_ESearch_)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def esearch(db, term):\n",
    "    \n",
    "    url    = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi\"\n",
    "    params = {\n",
    "                'db'        : db,\n",
    "                'term'      : term,\n",
    "                'usehistory': 'y',        # Important. Permet de revenir sur son historique de recherche sur NCBI\n",
    "                'RetMax'    : '25',\n",
    "                'retmode'   : 'json',\n",
    "                'sort'      : 'relevance'\n",
    "    }\n",
    "    return requests.get(url, params=params)\n",
    "    \n",
    "    \n",
    "    \n",
    "# 1. Interroger l'utilisateur\n",
    "kw = input('Quel gène cherchez-vous? Entrez un mot-clé (Ex. epo) : ')\n",
    "\n",
    "# 2. Interroger l'api\n",
    "term    = f\"human[orgn] (${kw}[gfn] OR ${kw}[gn] OR ${kw}[uid])\" # Construction des termes de la requête (v. documentation)\n",
    "req     = esearch(db='gene', term=term)                          # On appelle l'API\n",
    "results = req.json()                                             # Json parsing\n",
    "IDs     = results['esearchresult']['idlist']                     # Parmis les résultats, on veut la liste des identifiants\n",
    "\n",
    "# 3. Important de conserver si l'on veut utiliser l'historique de recherche sur NCBI...\n",
    "query_key = results['esearchresult']['querykey']\n",
    "WebEnv    = results['esearchresult']['webenv']\n",
    "\n",
    "# 4. Afficher; fermer\n",
    "print('\\nRésultats retournée :\\n', results)\n",
    "print('\\nListe des identifiants :\\n', IDs)\n",
    "print(\"\\nInfos pour l'historique :\", f'query_key : {query_key}', f'WebEnv    : {WebEnv}', sep='\\n')\n",
    "req.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ESummary\n",
    "\n",
    "On cherche une base de donnée (toujours la bd 'gene') à l'aide d'un identifiant cette fois. Permet d'obtenir beaucoup plus d'informations. Plus de détail [ici](https://www.ncbi.nlm.nih.gov/books/NBK25499/#_chapter4_ESummary_)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def esummary(db, id):\n",
    "\n",
    "    url    = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi\"\n",
    "    params = {\n",
    "                'db'     : db,\n",
    "                'id'     : id,\n",
    "                'retmode': 'json'\n",
    "             }\n",
    "    return requests.get(url, params=params)\n",
    "\n",
    "\n",
    "\n",
    "# 1. Interroger l'utilisateur\n",
    "id = input('Quel gène cherchez-vous? Entrez un id (Ex. 2056) : ')\n",
    "\n",
    "# 2. Interroger l'api\n",
    "req     = esummary(db='gene', id=id)\n",
    "results = req.json()\n",
    "print('\\nRésultats retournée :\\n', results)\n",
    "\n",
    "# 3. Sélectionner quelques infos et sauvegarder dans un dictionnaire\n",
    "fields  = ('uid', 'name', 'description', 'otheraliases', 'otherdesignations', 'summary', 'genomicinfo')\n",
    "id      = str(id.split(',')[0])   # Si plusieurs id, on ne garde que le premier (pour cet exemple)\n",
    "data    = {i:results['result'][id].get(i) for i in fields if results['result'].get(id)}   # Compréhension de dictionnaire\n",
    "print('\\nDictionnaire avec quelques infos sélectionnées :\\n', data)\n",
    "\n",
    "# 4. Afficher les résultats formatés\n",
    "print('\\nInfos formatées sur le gène : ')\n",
    "for k, v in data.items():\n",
    "    if k == 'genomicinfo' and v:\n",
    "        for k2, v2 in v[0].items(): print(k2.upper(), v2, sep=' : ')\n",
    "    else:\n",
    "        print(k.upper(), v, sep=' : ')\n",
    "req.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exemple avec l'utilisation de l'historique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def esummary2(db, query_key, WebEnv):\n",
    "\n",
    "    url    = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi\"\n",
    "    params = {\n",
    "                'db'       : db,\n",
    "                'retmode'  : 'json',\n",
    "                'query_key': query_key,\n",
    "                'WebEnv'   : WebEnv\n",
    "    }\n",
    "    return requests.get(url, params=params)\n",
    "\n",
    "\n",
    "# 1. Interroger l'utilisateur\n",
    "history              = dict()\n",
    "history['query_key'] = input(\"Entrez le query_key (Ex. 1) : \")\n",
    "history['WebEnv']    = input(\"Entrez le WebEnv (Ex. MCID_5f5e45e2d29...) : \")\n",
    "\n",
    "# 2. Interroger l'API\n",
    "req = esummary2(db='gene', **history)\n",
    "        \n",
    "# 3. Affichage\n",
    "print('\\nRésultats retournée :\\n', req.json()); req.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Elink\n",
    "\n",
    "Permet d'accéder aux références croisées (i.e. liens entre certains enregistrements d'une base de donnée à une autre).\n",
    "\n",
    "Par exemple, permet d'obtenir le lien vers la séquence d'ADN (bd 'nuclootide', aka 'nuccore') à partir de la fiche d'un gène (bd 'gene'). On utilise ce lien pour lier les enregistrements entre-eux.\n",
    "\n",
    "Liste des nombreux paramètres dans [Entrez Link Descriptions](https://eutils.ncbi.nlm.nih.gov/entrez/query/static/entrezlinks.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def elink(dbFrom, db, id, cmd='neighbor'):\n",
    "\n",
    "    url    = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi\"\n",
    "    params = {\n",
    "                'dbfrom'  : dbFrom,\n",
    "                'db'      : db,\n",
    "                'id'      : id,\n",
    "                'retmode' : 'json',\n",
    "                'cmd'     : cmd,\n",
    "                'linkname': 'gene_nuccore_refseqgene' # Existe de nombreux autres paramètres pour 'linkname'...\n",
    "    }\n",
    "    return requests.get(url, params=params)\n",
    "\n",
    "\n",
    "# 1. Interroger l'utilisateur\n",
    "id = input(\"Pour quel gène cherchez-vous la référence RefSeq? Entrez un id (Ex. 2056) : \")\n",
    "\n",
    "# 2a. Interroger l'API et affichage du lien RefSeq, sans l'historique\n",
    "# Lien RefSeq = Lien vers la séquence de référence\n",
    "# req    = elink(dbFrom='gene', db='nuccore', id=id)\n",
    "# refseq = req.json()['linksets'][0]['linksetdbs'][0]['links'][0]\n",
    "# print(f'\\nLien vers RefSeq : {refseq}')\n",
    "\n",
    "# 2b. Interroger l'API et affichage de l'historique, sans le lien RefSeq\n",
    "req       = elink(dbFrom='gene', db='nuccore', id=id, cmd='neighbor_history')\n",
    "query_key = req.json()['linksets'][0]['linksetdbhistories'][0]['querykey'][0]\n",
    "WebEnv    = req.json()['linksets'][0]['webenv']\n",
    "print(f'query_key : {query_key}', f'WebEnv    : {WebEnv}', sep='\\n')\n",
    "\n",
    "req.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Efetch\n",
    "\n",
    "On veut obtenir de l'information sur une séquence génique à l'aide d'un **numéro d'accession**. Plus de détail [ici](https://www.ncbi.nlm.nih.gov/books/NBK25499/#_chapter4_EFetch_). Aussi, si vous êtes intéressés par la [nomenclature](https://www.ncbi.nlm.nih.gov/Sequin/acc.html) des numéros d'accession..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fichier au format FASTA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def efetch(db, id):\n",
    "\n",
    "    url    = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "    params = {\n",
    "                'db'      : db,\n",
    "                'id'      : id,\n",
    "                'rettype' : 'fasta'\n",
    "    }\n",
    "    return requests.get(url, params=params)\n",
    "\n",
    "\n",
    "# Format FASTA\n",
    "accession = input(\"Quel séquence voulez-vous? Entrez un numéro d'accession (Ex. 'NG_021471.2') : \")\n",
    "req       = efetch(db='nuccore', id=accession)\n",
    "print('\\nRésultats retournés en format FASTA :\\n', req.text); req.close()\n",
    "\n",
    "# Être patient, ça peut être long..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fiche GenBank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def efetch2(db, id):\n",
    "\n",
    "    url    = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "    params = {\n",
    "                'db'     : db,\n",
    "                'id'     : id,\n",
    "                'rettype': 'gb'\n",
    "    }\n",
    "    return requests.get(url, params=params)\n",
    "\n",
    "\n",
    "# Obtenir une fiche GenBank\n",
    "accession = input(\"Quel séquence voulez-vous? Entrez un numéro d'accession (Ex. 'NG_021471.2') : \")\n",
    "req       = efetch2(db='nuccore', id=accession)\n",
    "print('\\nFiche GenBank :\\n\\n', req.text); req.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fiche GenBank avec *parsing* XML\n",
    "\n",
    "Réf. [ici](https://www.ncbi.nlm.nih.gov/dtd/NCBI_GBSeq.mod.dtd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from xml.etree import ElementTree # Pour le parsing du XML\n",
    "\n",
    "\n",
    "def efetch3(db, id):\n",
    "\n",
    "    url    = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "    params = {\n",
    "                'db'     : db,\n",
    "                'id'     : id,\n",
    "                'rettype': 'gb',\n",
    "                'retmode': 'xml'\n",
    "    }\n",
    "    return requests.get(url, params=params)\n",
    "\n",
    "\n",
    "# Obtenir une fiche GenBank en format xml\n",
    "accession = input(\"Quel séquence voulez-vous? Entrez un numéro d'accession (Ex. 'NG_021471.2') : \")\n",
    "req       = efetch3(db='nuccore', id=accession)\n",
    "\n",
    "# Afficher le fichier brute\n",
    "# print(req.text)\n",
    "\n",
    "\n",
    "root = ElementTree.fromstring(req.text)\n",
    "req.close()\n",
    "\n",
    "# Afficher un apperçu...\n",
    "# for child in root:\n",
    "#     for grandChild in child:\n",
    "#         print(grandChild.tag, grandChild.text, sep=' : ')\n",
    "\n",
    "# Afficher la séquence\n",
    "for i in root.iter('GBSeq_sequence'): print(i.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Wrapping it all together!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "\n",
    "def main():\n",
    "    \n",
    "    ####################################################################################\n",
    "    # 1. ESEARCH -- Recherche de gène par mot-clé. Obtention d'une liste d'identifiants\n",
    "    kw     = input('Quel gène cherchez-vous? Entrez un mot-clé : ')\n",
    "    term   = f\"human[orgn] (${kw}[gfn] OR ${kw}[gn] OR ${kw}[uid])\"\n",
    "    params = {\n",
    "                'db'        : 'gene',\n",
    "                'term'      : term,\n",
    "                'usehistory': 'y',\n",
    "                'RetMax'    : '25',\n",
    "                'retmode'   : 'json',\n",
    "                'sort'      : 'relevance'\n",
    "    }\n",
    "    req    = ncbiEntrez(eutils='esearch', params=params)\n",
    "    ids    = req.json()['esearchresult']['idlist']; req.close()\n",
    "    if not ids:\n",
    "        print('Aucun résultats')\n",
    "        return None # Une façon de quiter la fonction...\n",
    "    ####################################################################################\n",
    "\n",
    "\n",
    "    ####################################################################################\n",
    "    # 2. ESUMMARY -- Recherche les infos correspondant aux gènes de la liste\n",
    "    params = {\n",
    "                'db'     : 'gene',\n",
    "                'id'     : ','.join(ids),\n",
    "                'retmode': 'json'\n",
    "    }\n",
    "    req     = ncbiEntrez(eutils='esummary', params=params)\n",
    "    summary = req.json(); req.close()\n",
    "    ####################################################################################\n",
    "    \n",
    "    \n",
    "    ####################################################################################\n",
    "    # 3. Choix du gène parmis la liste\n",
    "    while True:\n",
    "        print('\\nListe des gènes trouvés : ')\n",
    "        for index, value in enumerate(ids):\n",
    "            print(index, summary['result'][value]['description'], sep='. ')\n",
    "        choix = input('\\nChoix : ')\n",
    "        if choix.isnumeric() and int(choix) in range(len(ids)): break\n",
    "    ####################################################################################\n",
    "            \n",
    "    \n",
    "    ####################################################################################\n",
    "    # 4. ELINK -- Obtention du lien vers RefSeq (indirectement via l'historique)\n",
    "    params = {\n",
    "                'dbFrom'  : 'gene',\n",
    "                'db'      : 'nuccore',\n",
    "                'id'      : ids[int(choix)],\n",
    "                'cmd'     : 'neighbor_history',\n",
    "                'retmode' : 'json',\n",
    "                'linkname': 'gene_nuccore_refseqgene'\n",
    "    }\n",
    "    req       = ncbiEntrez(eutils='elink', params=params)\n",
    "    query_key = req.json()['linksets'][0]['linksetdbhistories'][0]['querykey'][0]\n",
    "    WebEnv    = req.json()['linksets'][0]['webenv']\n",
    "    req.close()\n",
    "    ####################################################################################\n",
    "    \n",
    "    \n",
    "    ####################################################################################\n",
    "    # 5. EFETCH -- Obtention de la séquence ADN en format fasta\n",
    "    params = {\n",
    "                'db'       : 'nuccore',\n",
    "                'query_key': query_key,\n",
    "                'WebEnv'   : WebEnv,\n",
    "                'rettype'  : 'fasta'\n",
    "    }\n",
    "    req     = ncbiEntrez(eutils='efetch', params=params)\n",
    "    print('\\nFichier FASTA :\\n', req.text); req.close()\n",
    "    ####################################################################################\n",
    "    \n",
    "\n",
    "    \n",
    "def ncbiEntrez(eutils, params, method='get'):\n",
    "    \n",
    "    url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/{eutils}.fcgi\"\n",
    "    return getattr(requests, method)(url, params=params) # 'getattr' appelle dynamiquement une méthode sur un objet.\n",
    "    \n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
